### PR TITLE
Uidreference backreferences

### PR DESCRIPTION
**Pull request by rockfruit**
```python
def get_backreferences(context, relationship=None, as_brains=None):
    """Return all objects which use a UIDReferenceField to reference context.

    :param context: The object which is the target of references.
    :param relationship: The relationship name of the UIDReferenceField.
    :param as_brains: Requests that this function returns only catalog brains.
        as_brains can only be used if a relationship has been specified.

    This function can be called with or without specifying a relationship.

    - If a relationship is provided, the return value will be a list of items
      which reference the context using the provided relationship.

      If relationship is provided, then you can request that the backrefs
      should be returned as catalog brains.  If you do not specify as_brains,
      the raw list of UIDs will be returned.

    - If the relationship is not provided, then the entire set of
      backreferences to the context object is returned (by reference) as a
      dictionary.  This value can then be modified in-place, to edit the stored
      backreferences.
    """
```

### CODE REVIEW

File: bika/lims/browser/fields/uidreferencefield.py [hunk 1]
```diff
@@ -207,5 +179,121 @@ def set(self, context, value, **kwargs):
...
+      return type.
+    """
+
+    instance = context.aq_base
+    backrefs = get_storage(instance)
+
+    if relationship:
+        backrefs = backrefs.get(relationship, [])
+        if not backrefs:
+            return backrefs
```
Comment (ramonski) [inline bika/lims/browser/fields/uidreferencefield.py]: Check for `if backrefs is None` and return an empty list `[]`

File: bika/lims/browser/fields/uidreferencefield.py [hunk 2]
```diff
@@ -207,5 +179,121 @@ def set(self, context, value, **kwargs):
...
+        # return only UIDs
+        if as_uids:
+            return backrefs
+        # return full objects
+        elif as_objects:
+            return [_get_object(context, b) for b in backrefs]
+        # Return brains from first catalog found.
+        cat = _get_catalog_for_uid(backrefs[0])
+        return cat(UID=backrefs)
```
Comment (ramonski) [inline bika/lims/browser/fields/uidreferencefield.py]: This logic seems to me confusing. Either we should return UIDs or objects depending on the parameters the caller of this function provided.

Better refactor this function as follows:

```
instance = context.aq_base
storage = get_storage(instance)
if relationship is None:
    assert not as_objects, "You cannot use as_objects with no relationship"
    return storage
backrefs = backrefs.get(relationship, None)
if backrefs in None:
    return []
if as_objects:
    return [_get_object(context, b) for b in backrefs]
return backrefs
```

File: bika/lims/browser/fields/uidreferencefield.py [hunk 3]
```diff
@@ -207,5 +179,121 @@ def set(self, context, value, **kwargs):
...
+
+    assert not as_objects, "You cannot use as_objects with no relationship"
+
+    return backrefs
+
+
+def _get_catalog_for_uid(uid):
```
Comment (ramonski) [inline bika/lims/browser/fields/uidreferencefield.py]: I think we won't need that function if we drop the assumption that the user eventually want's to have brains.

File: bika/lims/content/calculation.py [hunk 4]
```diff
@@ -245,8 +244,8 @@ def getCalculationDependants(self):
         calculation's Formula.
         """
         deps = []
-        for service_uid in self.service_backreferences:
-            service = get_object_by_uid(service_uid)
+        backrefs = get_backreferences(self, 'AnalysisServiceCalculation', as_objects=1)
```
Comment (ramonski) [inline bika/lims/content/calculation.py]: I would drop this `as_objects` completely and let the field always return UIDs. With a set of UIDs all caller functions can wake up the objects if they need it with `map(api.get_object_by_uid, get_backreferences(self, 'AnalysisServiceCalculation'))` easily

**Comment by rockfruit** [comment]
@ramonski:
> Please change the UIDReference field to always return a single UID or a list of UIDs and let the caller decide how to proceed with these UIDs. This will shift the data responsibility to the caller and reduces therefore complexity.

I agree that this will reduce the complexity of the backreferences code, however shifting this responsibility onto the caller will increase complexity in many places, and it will cause different callers, in different contexts, to approach the issue of retrieving brains in different ways.  For example, it is not the caller's responsibility to know which catalog an object is indexed in (and sometimes, can be multiple catalogs, at least at the moment this is true).
 
I suggest a compromise:  I will return UIDs by default (better than brains, because much faster), and I will remove any option to receive full objects - if full objects are required, then calling .getObject() or doing a lookup will remain the responsibility of the caller.  But I would like to continue to provide an `as_brains=None` to the signature, this is a convenience that I think is meaningful to callers.

File: bika/lims/browser/fields/uidreferencefield.py [hunk 5]
```diff
@@ -80,32 +45,20 @@ class UIDReferenceField(StringField):
...
-                # getSite doesn't fuck up.
-                uc = get_tool('uid_catalog')
-            brains = uc(UID=value)
-            if brains:
-                return brains[0].getObject()
-            logger.error(
-                "{}.{}: Resolving UIDReference failed for {}.  No object will "
-                "be returned.".format(context, self.getName(), value))
+        obj = _get_object(context, value)
+        if obj:
```
[the remark this answers is not in the document]
Comment (rockfruit) [reply bika/lims/browser/fields/uidreferencefield.py]: agree much better.

File: bika/lims/browser/fields/uidreferencefield.py [hunk 6]
```diff
@@ -147,9 +100,8 @@ def get(self, context, **kwargs):
         value = StringField.get(self, context, **kwargs)
```
[the remark this answers is not in the document]
Comment (rockfruit) [reply bika/lims/browser/fields/uidreferencefield.py]: also L108.

File: bika/lims/browser/fields/uidreferencefield.py [hunk 7]
```diff
@@ -207,5 +179,121 @@ def set(self, context, value, **kwargs):
...
+    :type context: BaseContent
+    :param value: A UID.
+    :type value: string
+    :return: Returns a Content object.
+    :rtype: BaseContent
+    """
+
+    if value:
+        if is_at_content(value) or is_dexterity_content(value):
+            return value
```
[the remark this answers is not in the document]
Comment (rockfruit) [reply bika/lims/browser/fields/uidreferencefield.py]: ta

File: bika/lims/browser/fields/uidreferencefield.py [hunk 8]
```diff
@@ -207,5 +179,121 @@ def set(self, context, value, **kwargs):
...
+        else:
+            uc = getToolByName(context, 'uid_catalog')
+            brains = uc(UID=value)
+            if brains:
+                return brains[0].getObject()
```
[the remark this answers is not in the document]
Comment (rockfruit) [reply bika/lims/browser/fields/uidreferencefield.py]: done.  I actually went blithly ahead and did this thinking "shouldn't happen".  my bad

File: bika/lims/browser/fields/uidreferencefield.py [hunk 9]
```diff
@@ -207,5 +179,121 @@ def set(self, context, value, **kwargs):
...
+
+      If relationship is not provided, the returned backreferences will
+      always be returned as UIDs, and it is an error to specify another
+      return type.
+    """
+
+    instance = context.aq_base
+    backrefs = get_storage(instance)
+
+    if relationship:
```
[the remark this answers is not in the document]
Comment (rockfruit) [reply bika/lims/browser/fields/uidreferencefield.py]: I have a belief that a function should have a single point of return, tho i can see this is better

**Pull request merged**